### PR TITLE
marquee text

### PR DESCRIPTION
**Pull request by Decimate1405**
Implemented marquee text in song menu for overflow, instead of two lines

### CODE REVIEW

**Comment by Chaphasilor** [comment]
Hey, thanks for the PR. Sorry for not responding sooner, but I'll take a look now.  
I'm not a huge fan of scrolling text myself but I'm happy with making this configurable via a setting.  
Is there a particular reason why you didn't use an existing package, like [`marquee`](https://pub.dev/packages/marquee)? There is also an existing solution (without a PR so far) at https://github.com/Ivanf1/finamp/tree/marquee that implements this for the player screen as a fallback if the two lines are not enough.

If you could update this PR to behave in a similar way (using up to two lines by default, and then switching to scrolling text) that would be nice. If you don't like the line break and would like to always keep it at a single line, that could be added through a setting!

Also, it seems like you removed the track title on the player screen?

**Comment by Decimate1405** [comment]
Hey, thank you for the response. I tried using marquee package but it was giving an error during import (some conflict with fading_edge_scrollview package) that's why I implemented it myself. I think adding it as a setting would be really nice as it would accommodate for everyone's needs. Regarding the title being removed from the player screen, looks like I accidentally forgot to remove the comment.

**Comment by Decimate1405** [comment]
added a toggle to switch between 1 or 2 lines of song title, defaults to 2. its under Player settings.

**Comment by Chaphasilor** [comment]
Okay, thanks for the changes. There's still a lot of commented-out code, that should be removed.  
Could you also try integrating the `marquee` package again, instead of using the custom implementation? Or is there a specific benefit to your implementation?

**Comment by Decimate1405** [comment]
There's no benefit to using my implementation. I couldn't fix the errors i was getting while using the marquee package so i got fed up and implemented it myself. I'll give it another go

**Comment by Decimate1405** [comment]
This is the error I keep getting when I use the marquee package in the now playing bar.
Screenshot 2024-05-31 at 2 47 03 PM

**Comment by Chaphasilor** [comment]
Try adding the override from https://github.com/MarcelGarus/marquee/issues/98#issuecomment-2119314341 in pubspec.yaml :)

**Comment by Decimate1405** [comment]
changed it to use the marquee package, still defaults to 2 lines

**Comment by Chaphasilor** [comment]
Appreciate the effort! Those changes sound good.

Let me know when I should try this out!

**Comment by Decimate1405** [comment]
its working, you can try it now :)

**Comment by Chaphasilor** [comment]
Okay, took a look at the changes just now, here's what I noticed:

Marquee text color on the now playing bar doesn't adapt to the theme:
image

If a marquee isn't needed, the fallback isn't using the `BalancedText` widget:
image
Here's what it should look like:
image

With the one line marquee enabled, even non-overflowing text uses a marquee on the player screen:
image

The marquee should probably also be used in the queue list:
image
(code is very similar, so it should be easy to add)

And maybe also in the track menu:
image

And finally, one suggestion. Do you think you could add a mask to the marquee on the player screen, so that it fades out at the sides instead of simply being cut off? You can take a look at the headers within the queue list as an example. Here's what I mean:
image

Overall, the change is nice, and once the kinks are ironed out, I believe this will be an awesome new feature!  
I'll probably add a commit that updated the settings title and descriptions to be a bit clearer, but that's it :)

**Comment by Decimate1405** [comment]
thanks for the feedback. i'll work on the changes

**Comment by Decimate1405** [comment]
made all the changes mentioned:
1. Marquee text color on the now playing bar adapts to the theme
Screenshot 2024-06-03 at 7 23 03 AM

2. If a marquee isn't needed, the fallback uses the BalancedText widget
Screenshot 2024-06-03 at 7 24 08 AM

3. Fixed the bug where with the one line marquee enabled, even non-overflowing text uses a marquee on the player screen
Screenshot 2024-06-03 at 7 29 15 AM

4. Marquee added in queue list
Screenshot 2024-06-03 at 7 27 15 AM

5. Marquee added in track/album menu
Screenshot 2024-06-03 at 7 28 04 AM

**Comment by Chaphasilor** [comment]
Taking a look now, I missed your comment, sorry about that.

By "track menu", I meant the menu that opens when tapping the "hamburger icon" or three dot icon, with options like "add to playlist", "go to artist", etc. From your screenshot it seems like you added the marquee to the track list within albums and playlists, which is probably a bit overkill...

**Comment by Decimate1405** [comment]
oh, I thought you meant track/album menu which did seem a bit off to me. i can move it to the hamburger menu

**Comment by Chaphasilor** [comment]
Also, the text color on the now playing bar is still wrong in light mode. It should be white there as well, as before. So in a way you could say it should *not* adapt to the theme :P

image

**Comment by Chaphasilor** [comment]
I also think there's no need to add marquees for all the other tracks in the queue list. Just the currently playing track should be enough, otherwise there's too much movement on the screen. But I'd like to hear your opinion.

**Comment by Decimate1405** [comment]
i do agree, when i was testing it out having marquee in the queue list seemed very off because none of the tracks are synced because of their title length.

**Comment by Decimate1405** [comment]
Just to confirm, this text should always be white regardless of system theme right?
Dark theme
Screenshot 2024-06-04 at 5 49 17 PM
Screenshot 2024-06-04 at 5 50 40 PM


Light Theme
Screenshot 2024-06-04 at 5 49 52 PM

Screenshot 2024-06-04 at 5 50 18 PM

**Comment by Chaphasilor** [comment]
Yes. That's how it should be :)

I also noticed that the height required by marquee text and two-line wrapped text isn't equal. Previously I made sure that one line and two line titles had the same height (single line was centered vertically), so that so layout shift between tracks happens.  
Can you make the marquee on the player screen have the same height?

(tried to record a video of the problem, but my screen recorder doesn't like me)

**Comment by Decimate1405** [comment]
oh ok. yea I intentionally changed the height between one line and two line because i thought it would look cleaner. but i can keep it the same

**Comment by Chaphasilor** [comment]
Also, please add a "Scroll text instead of truncating" setting to the customization settings at, it can default to be enabled. You might also wanna move the other setting over there, since it doesn't only affect the player screen anymore :)

**Comment by Chaphasilor** [comment]
Okay, here's the layout shift I was talking about:

https://github.com/jmshrv/finamp/assets/18015147/7bf035cc-33bc-4082-99dd-0d6b59cd4a37

**Comment by Decimate1405** [comment]
added marquee in hamburger menu track title (kept the same text style as before)
Screenshot 2024-06-04 at 6 03 46 PM



> Also, please add a "Scroll text instead of truncating" setting to the customization settings at, it can default to be enabled. You might also wanna move the other setting over there, since it doesn't only affect the player screen anymore :)

do you mean I rename "One Line Marquee" to "Scroll text instead of truncating" and move from Player screen settings to Customization settings?

**Comment by Decimate1405** [comment]
is this layout fix good or should i change something?
[Screen_recording_20240604_181544.webm](https://github.com/jmshrv/finamp/assets/24218060/786b08a0-4d22-4359-9a8c-b62328b41cea)

**Comment by Chaphasilor** [comment]
> do you mean I rename "One Line Marquee" to "Scroll text instead of truncating" and move from Player screen settings to Customization settings?

No, add a new setting that controls if a marquee is used at all (or if the previous behavior with the ellipsis is always used) and move the other setting

Or you could try to combine those two settings in a select/dropdown list.  
That could be "Handle too-long text by..." and then three options: "truncating" (old behavior), "scrolling" (one-line marquee), "wrapping and scrolling" (two-line marquee).  
But I'm not sure if that's really a good idea, the two separate settings seem easier for now :)

**Comment by Chaphasilor** [comment]
Layout change is almost perfect, but the marquee text should be at the same position as the non-marquee single-line text. Here you can see they are not quite at the same height:

**Comment by Chaphasilor** [comment]
@Decimate1405 are you still willing to work on this? I can handle the merge conflict for you, if that helps?

**Comment by Decimate1405** [comment]
Yes, last few weeks have been quite busy but i can continue working on it this weekend. Could you just summarize what other features are needed. The last change I made was the layout fix

**Comment by Chaphasilor** [comment]
That's great to hear!  
The changes I'd love to see are:
- Make sure marquee text looks the same (font, vertical position, etc.) as a single-line title (i.e. [this comment](https://github.com/jmshrv/finamp/pull/754#issuecomment-2148542453))
- Add a new toggle in customization settings that controls if marquees should be used at all (can default to true). If disabled, overflowing text should be ellipsized like before
- Move the existing "one line marquee" setting to the customization settings. That only has an effect if the previous setting is enabled. Maybe reword it as "allow one line break before using scrolling text" or something similar

Hopefully nothing major :)

**Comment by Chaphasilor** [comment]
Hey, just saw you added commits to this. Is this ready to re-review? If so, would you mind resolving the merge conflicts? :)

**Comment by Chaphasilor** [comment]
@Decimate1405 I can resolve the conflicts and merge the PRs if you're busy. Just need your OK!

**Comment by Decimate1405** [comment]
Hey @Chaphasilor ! Sorry I have been busy with things in personal life that I haven't been able to complete it, I still have to implement the default ellipses functionality and fix the centering a bit. I can complete it in the next 2 weeks and try to solve the merge conflicts as well

**Comment by Chaphasilor** [comment]
@Decimate1405 no worries! I just wanted to make sure the work you put into this will end up being used 😁  
If you can finish it (at some point) that would be nice, but if you can't get to it, just let me know ^^

**Comment by Decimate1405** [comment]
Hey @Chaphasilor! It's been longer than I said it would take 😅, but I made the changes mentioned:

1. Marquee text formatting is same whether title is single or double lines
image
image
image
2. Added a toggle in customization menu that will allow user to force truncate title in player screen if they don't like marquee, both the settings can be toggled on, this will make the title in miniplayer use the marquee effect while truncating it in the player screen
image

Let me know if any more changes are needed, i'll fix it immediately 🫡

**Comment by Decimate1405** [comment]
Yea sure I can make those changes. For the descriptions would this be alright or should I simplify it further?
{
  "oneLineMarqueeTextButton": "Single line text display",
  "oneLineMarqueeTextButtonSubtitle": "Show song titles in a single line instead of two lines",
  
  "marqueeOrTruncateButton": "Use ellipsis for long titles",
  "marqueeOrTruncateButtonSubtitle": "Show ... at the end of long titles instead of scrolling text"
}

I'll give the long artist / album names a try

**Comment by Decimate1405** [comment]
Could you show an example of the problem or redirect me to wherever this issue has been discussed so I can get a better idea of what I need to implement? Right now all the long name albums I have fit perfectly in the smallest screen (4.6") I can emulate so I'm unable to reproduce it.

**Comment by Chaphasilor** [comment]
Okay, just from your comments so far I have some feedback:
1. I assumed that with the `oneLineMarquee` setting, the marquee would be "always enabled", while with the setting disabled the marquee would only be used if the text still doesn't fit on two lines. Your new button title makes it sound like there will be no marquee at all when the setting is disabled?  
The new wording for the other setting is nice!
2.  The "chip" for the album has more padding than before, from looking at your screenshots. But it should look like before (at least if no marquee is needed).
3. Can you try to limit the maximum width of the "chip" (or the container around it) so that it won't grow larger than the artist "chip"? Basically, take a look at the two buttons (favorite and hamburger menu) next to the artist. They should limit the size of all three info lines, the title, the artists, and the album. The space directly above and below these buttons should be empty.  
This also means that you can make the title marquee a bit wider. Take a look at the screenshots you posted in https://github.com/jmshrv/finamp/pull/754#issuecomment-2621445402, the marquee start position (where the "b" of "probably" is) should be at the same distance from the left as the edge of the album chip)

I hope that is clear enough, if not I'll try to explain it again!

**Comment by Decimate1405** [comment]
ok so I have implemented that big album name issue
image

but I had a question for the marquee one, so right now i have two toggles:
1: switches marquee on/off - if its on then anything more than 1 line is marquee, if off then anything above 2 lines is marquee
2: switches truncate on/off - if on any number of lines if overflowed is truncated by ...

now that I look back at this I'm not even sure why I broke it into two settings. so the question is do we need these two different settings or should I just have marquee on/off toggle where on would make anything more than 1 line into marquee and and off would truncate anything that overflows.

**Comment by Chaphasilor** [comment]
I suggested the two line option. Especially on the player screen we often have enough space to statically show the entire title, but if that doesn't fit, we need to truncate or use marquee.  
So the two settings aren't about on/off, but about *when* (how soon) overflow is handled, and *how* it is handled. Does that clear it up?

I can't take a proper look at the PR atm since my server died today, but hopefully I can get things up and running soon...

**Comment by Decimate1405** [comment]
I'm not sure I understand it completely

> but if that doesn't fit, we need to truncate or use marquee

so this essentially what marquee is supposed to do right? like we give the user the option whether they wanna use marquee or not, and if they choose not to then the overflow is handled by truncating it. so the "on/off" is basically the user toggling between whether they want marquee or not (truncate)

if im wrong then yea elaborating a bit more will be helpful

**Comment by Decimate1405** [comment]
so this is how currently this toggle works

1. off state
image

image

image

2. on state
image

image

image

**Comment by Chaphasilor** [comment]
@Decimate1405 what I meant is that even when the setting is on, the title should still be shown in two lines as long as it doesn't overflow. Only if the setting is enabled *and* the title is longer than two lines should the title be shown as a one line marquee.

As I said, I'm currently not able to take a look at this myself since my server is broken and I don't have any media downloaded for the Finamp debug version. Otherwise I could just tell you if it behaves as expected...

**Comment by Decimate1405** [comment]
Ok I understood what you are trying to say. The latest commit should be the last commit unless some more changes are required. Take a look at it whenever your server is back.

**Comment by Chaphasilor** [comment]
@Decimate1405 just took a look, yes the new setting is all that's needed I think, good job 👍🏻  
The album chip wasn't quite what I had in mind, instead of a marquee I'd just have it be manually scrollable, like when having multiple artists or feature chips. Do you have the time to adjust that? If not, I can do it myself :)

**Comment by Chaphasilor** [comment]
@Decimate1405 I've changed the album chip (and artist chip) behavior to be scrollable instead of using a marquee, which is what I meant. This is more in line with how scrolling for multiple artists behaves. The marquee is still used in any cases where manual scrolling is not possible :)

Once the build checks are complete I'll go ahead and finally merge this, so it ends up in the next beta update!
